caching/base.py

```diff
@@ -157,6 +157,18 @@ def __setstate__(self, state):
         if self.timeout == self._default_timeout_pickle_key:
             self.timeout = DEFAULT_TIMEOUT
 
+    def _fetch_all(self):
+        """
+        Django 1.11 changed _fetch_all to use self._iterable_class() rather than
+        self.iterator(). That bypasses our iterator, so override Queryset._fetch_all
+        to use our iterator.
+
+        https://github.com/django/django/commit/f3b7c059367a4e82bbfc7e4f0d42b10975e79f0c#diff-5b0dda5eb9a242c15879dc9cd2121379
+        """
+        if self._result_cache is None:
+            self._result_cache = list(self.iterator())
```

Comment:
I believe this change will cause a server-side cursor to be used for `_fetch_all` (if not globally disabled via `DISABLE_SERVER_SIDE_CURSORS`) on Django 1.11.1+ due two these lines:

- https://github.com/vkurup/django-cache-machine/blob/6a547d8e4295985b2a84d4e115b1192cff540dbc/caching/base.py#L181
- https://github.com/django/django/blob/stable/1.11.x/django/db/models/query.py#L322-L323

I think we need to make sure `_fetch_all()` still uses `self._iterable_class()` directly rather than `self.iterator()` on Django 1.11+, perhaps via something along the lines of the following:

- Separate our `iterator()` logic into a separate method, e.g., `_iterate_on(iterable)`
- In `_fetch_all()`, call `list(self._iterate_on(iter(self._iterable_class(self))))` for Django 1.11+ or `list(self._iterate_on(self.iterator()))` for older Djangos (`_iterable_class` unfortunately does not exist in Django 1.8).
- In `iterator()`, simply return `self._iterate_on(super(CachingQuerySet, self).iterator())`

Does that seem crazy? What do you think?

Reply:
Or better yet, rather than check the version directly, use duck typing as you already did here: https://github.com/django-cache-machine/django-cache-machine/pull/125/commits/24c977923ab5fee5525fb5fea0df6b4859e53316#diff-76150a51cd21d8d6b30aaaab22c7f114R179

`_fetch_all()` and `iterator()` are unchanged between Django 1.9-1.10, so it should be safe to use `self._iterable_class` directly in those versions (in addition to Django 1.11)

`_fetch_all()`:
- 1.9: https://github.com/django/django/blob/stable/1.9.x/django/db/models/query.py#L1072-L1076
- 1.10: https://github.com/django/django/blob/stable/1.10.x/django/db/models/query.py#L1085-L1089

`iterator()`:
- 1.9: https://github.com/django/django/blob/stable/1.9.x/django/db/models/query.py#L325-L330
- 1.10: https://github.com/django/django/blob/stable/1.10.x/django/db/models/query.py#L323-L328